### PR TITLE
Geolocator

### PR DESCRIPTION
**Pull request by lekshmanmj**
**PR on GPS Location Feature**
@FeodorFitsner @ndonkoHenri 

As Stream function provided in Geolocator dart package just stops streaming when app is minimized, Initially I planned to implement location streaming using Background Service... But while trying that in a clean/pure flutter project itself I got stuck with some Kotlin version related error(which I couldn't solve). Therefore I finally decided to just have a normal function to get location in dart side and decided to call it explicitly in a loop in flet app using run_task/run_thread method to implement stream(to leave it to user's/developer's side) as implemented in the code.

playground/main.py
```python
import flet as ft
import asyncio


def main(page: ft.Page):
    page.scroll = ft.ScrollMode.ADAPTIVE
    page.add(ft.SafeArea(ft.Text("Geolocator test")))
    gl = ft.Geolocator()
    page.overlay.append(gl)
    page.update()
    position = gl.get_location()
    page.add(
        ft.SafeArea(
            ft.Text(
                "latitude: {}, longitude: {}".format(
                    position.latitude, position.longitude
                )
            )
        )
    )
    page.add(ft.SafeArea(ft.Text("Geolocation Stream using Thread/AsyncTask")))

    async def loop():
        while True:
            position = await gl.get_location_async()
            page.add(
                ft.SafeArea(
                    ft.Text(
                        "latitude: {}, longitude: {}".format(
                            position.latitude, position.longitude
                        )
                    )
                )
            )
            await asyncio.sleep(1)

    page.run_task(loop)
    page.update()


ft.app(main)

```

### CODE REVIEW

**Comment by FeodorFitsner** [comment]
Thank you! Useful control - great PR! 👍

I'm looking at the example for Geolocator widget: https://pub.dev/packages/geolocator#example

I think we need to implement methods to check/request permissions:
* `isLocationServiceEnabled()`
* `checkPermission()`
* `requestPermission()`
* `LocationPermission` enum.

We obviously can't rely on an assumption the location service is allowed/enabled on user's device, right?
What do you think?

File: packages/flet_geolocator/lib/src/geolocator.dart [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      case "high":
+        locationAccuracy = LocationAccuracy.high;
+      case "medium":
+        locationAccuracy = LocationAccuracy.medium;
+      case "low":
+        locationAccuracy = LocationAccuracy.low;
+      case "lowest":
+        locationAccuracy = LocationAccuracy.lowest;
+      case "reduced":
+        locationAccuracy = LocationAccuracy.reduced;
```
Comment (ndonkoHenri) [inline packages/flet_geolocator/lib/src/geolocator.dart:66]: Please move this to a utils file as seen here: https://github.com/flet-dev/flet/pull/3093/files#diff-2c856694204af9aa4aa908b108a4541b8dea385b5bb1f57870db9cdcc716c945

Then use it as seen here: https://github.com/flet-dev/flet/pull/3093/files#diff-a0b7d8d011def02618f1813be955bf63ad0b894bb238cb50c3087941d60027d8R54

File: packages/flet_geolocator/pubspec.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+name: flet_geolocator
+description: Flet Geolocator control
+homepage: https://flet.dev
+repository: https://github.com/flet-dev/flet/packages/flet_geolocator
+version: 0.22.0
```
Comment (ndonkoHenri) [inline packages/flet_geolocator/pubspec.yaml:5]:
```suggestion
version: 0.23.0
```

Your addition will most probably be part of Flet v0.23.0, so let's modify.

File: sdk/python/packages/flet-core/src/flet_core/geolocator.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+from flet_core.ref import Ref
+
+
+class LocationAccuracy(Enum):
+    LOWEST = "lowest"
+    LOW = "low"
+    MEDIUM = "medium"
+    HIGH = "high"
+    BEST = "best"
+    BESTFORNAVIGATION = "bestForNavigation"
```
Comment (ndonkoHenri) [inline sdk/python/packages/flet-core/src/flet_core/geolocator.py:14]: Let's separate this:
```suggestion
    BEST_FOR_NAVIGATION = "bestForNavigation"
```

File: packages/flet_geolocator/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+No changes in this release. Version bumped to follow parent `flet` package.
+
+# 0.20.1
+
+No changes in this release. Version bumped to follow parent `flet` package.
+
+# 0.20.0
+
+Initial release of the package.
```
Comment (ndonkoHenri) [inline packages/flet_geolocator/CHANGELOG.md:27]: Please cleanup here (remove all the useless versions/descriptions) leaving only the below suggestion:
```suggestion
# 0.23.0

Initial release of the package.
```

File: packages/flet_geolocator/lib/src/geolocator.dart [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+              if (value == true) {
+                Position location = await Geolocator.getCurrentPosition(
+                    desiredAccuracy: locationAccuracy);
+
+                return 'latitude.${location.latitude},longitude.${location.longitude},altitude.${location.altitude},speed.${location.speed},timestamp.${location.timestamp}';
+              } else {
+
+                return 'latitude null,longitude null,altitude null,speed null,timestamp null';
+              }
```
Comment (ndonkoHenri) [inline packages/flet_geolocator/lib/src/geolocator.dart:86]: We will need to json-encode this before sending it over to python end for decoding. 
Will look for an example illustrating how it is done, and share it here.

**Comment by lekshmanmj** [comment]
> Thank you! Useful control - great PR! 👍
> 
> I'm looking at the example for Geolocator widget: https://pub.dev/packages/geolocator#example
> 
> I think we need to implement methods to check/request permissions:
> 
> * `isLocationServiceEnabled()`
> * `checkPermission()`
> * `requestPermission()`
> * `LocationPermission` enum.
> 
> We obviously can't rely on an assumption the location service is allowed/enabled on user's device, right? What do you think?

`Permission.location.request().isGranted` This line does the job..
way 1 - if Permission is not given : It prompts the user for both permission and to enable location service --> waits infinitely to get users response --> return true/false
way 2- If Permission is already granted: It just directly return true/false.
I have witnessed the above scenario while test in clean flutter project in my android device.

I ignorned package(geolocator) provided permission related function in current implementation, as I have better understanding about common PermissionHandler dart package.

Planned to add `check_permission` and `request_permission` functions to common flet_permission_handler feature(I will submit a PR on it shortly).

File: packages/flet_geolocator/lib/src/geolocator.dart (hunk 1 above)
Comment (lekshmanmj) [reply packages/flet_geolocator/lib/src/geolocator.dart:66]: done

File: packages/flet_geolocator/pubspec.yaml (hunk 2 above)
Comment (lekshmanmj) [reply packages/flet_geolocator/pubspec.yaml:5]: done

File: sdk/python/packages/flet-core/src/flet_core/geolocator.py (hunk 3 above)
Comment (lekshmanmj) [reply sdk/python/packages/flet-core/src/flet_core/geolocator.py:14]: done

File: packages/flet_geolocator/CHANGELOG.md (hunk 4 above)
Comment (lekshmanmj) [reply packages/flet_geolocator/CHANGELOG.md:27]: done

File: packages/flet_geolocator/lib/src/geolocator.dart (hunk 5 above)
Comment (lekshmanmj) [reply packages/flet_geolocator/lib/src/geolocator.dart:86]: in flet_control_dart.dart 
image

return type is Future<String?> output can't be returned in json  format now.

**Comment by ndonkoHenri** [comment]
> I've ignorned package(geolocator) provided permission related function in current implementation, as I have better understanding about common PermissionHandler dart package.

Seems like the `permissions_handler` package is [limited](https://pub.dev/packages/permission_handler#:~:text=provides%20a%20cross%2Dplatform-,(iOS%2C%20Android),-API%20to%20request%20permissions) to mobile only? 😯

In any case, I suggest we go for independence: expose all the possible methods provided by the `geolocator` package as they might be optimised for its specific use-case. 
By the way, seems like geolocator does [not](https://pub.dev/packages?q=dependency%3Apermission_handler+geolocator) depend on permissions_handler.

`permission_handler` package you [implemented](https://github.com/flet-dev/flet/pull/3212), will of course still be useful - perhaps not that much for the location-related checks, but for all others. (thanks for working on it!)

**Comment by lekshmanmj** [comment]
> > I've ignorned package(geolocator) provided permission related function in current implementation, as I have better understanding about common PermissionHandler dart package.
> 
> Seems like the `permissions_handler` package is [limited](https://pub.dev/packages/permission_handler#:~:text=provides%20a%20cross%2Dplatform-,(iOS%2C%20Android),-API%20to%20request%20permissions) to mobile only? 😯
> 
> In any case, I suggest we go for independence: expose all the possible methods provided by the `geolocator` package as they might be optimised for its specific use-case. By the way, seems like geolocator does [not](https://pub.dev/packages?q=dependency%3Apermission_handler+geolocator) depend on permissions_handler.
> 
> `permission_handler` package you [implemented](https://github.com/flet-dev/flet/pull/3212), will of course still be useful - perhaps not that much for the location-related checks, but for all others. (thanks for working on it!)

image

yes... will expose all other functions of that package.

Geolocator is directly dependent on permission_handler.dart rather than on permission_handler which is recently implemented in flet.

permission_handler is intended to be used in rare scenarios only... like if user wish to check/get multiple permission on opening app.

**Comment by ndonkoHenri** [comment]
So, I went on and added/updated some little stuffs such as permissions, utils and the exposed methods... 
I think your PR is now ready for it's v1. Additionally, if you have desire and time, you can pick up from here and expose more stuffs... (no problem when you can't - I could do that some other day).

I tested simultaneously on 4 platforms and it worked: android, mac, web and ios - only windows is left out now. 
image

### Updated Test Code
```py
import flet as ft


async def main(page: ft.Page):
    page.window_always_on_top = True
    page.on_error = lambda e: print(f"Page Error: {e.data}")
    page.scroll = ft.ScrollMode.ADAPTIVE
    page.appbar = ft.AppBar(title=ft.Text("Geolocator Tests"))
    gl = ft.Geolocator()
    page.overlay.append(gl)

    settings_dlg = lambda handler: ft.AlertDialog(
        adaptive=True,
        title=ft.Text("Opening Location Settings..."),
        content=ft.Text(
            "You are about to be redirected to the location/app settings. "
            "Please locate this app and grant it location permissions."
        ),
        actions=[
            ft.TextButton(
                text="OK",
                on_click=handler,
            ),
        ],
        actions_alignment=ft.MainAxisAlignment.CENTER,
    )

    def handle_permission_request(e):
        page.add(ft.Text(f"request_permission: {gl.request_permission()}"))

    def handle_has_permission(e):
        page.add(ft.Text(f"has_permission: {gl.has_permission()}"))

    def handle_get_current_position(e):
        p = gl.get_current_position()
        page.add(ft.Text(f"get_current_position: ({p.latitude}, {p.longitude})"))

    def handle_get_last_known_position(e):
        p = gl.get_last_known_position()
        page.add(ft.Text(f"get_last_known_position: ({p.latitude}, {p.longitude})"))

    def handle_location_service_enabled(e):
        page.add(
            ft.Text(f"is_location_service_enabled: {gl.is_location_service_enabled()}")
        )

    def handle_open_location_settings(e):
        page.close_dialog()
        page.add(ft.Text(f"open_location_settings: {gl.open_location_settings()}"))

    def handle_open_app_settings(e):
        page.close_dialog()
        page.add(ft.Text(f"open_app_settings: {gl.open_app_settings()}"))

    page.add(
        ft.Row(
            [
                ft.OutlinedButton(
                    "request_permission",
                    on_click=handle_permission_request,
                ),
                ft.OutlinedButton(
                    "has_permission",
                    on_click=handle_has_permission,
                ),
                ft.OutlinedButton(
                    "get_current_position",
                    on_click=handle_get_current_position,
                ),
                ft.OutlinedButton(
                    "get_last_known_position",
                    visible=False if page.web else True,
                    on_click=handle_get_last_known_position,
                ),
                ft.OutlinedButton(
                    "is_location_service_enabled",
                    on_click=handle_location_service_enabled,
                ),
                ft.OutlinedButton(
                    "open_location_settings",
                    visible=False if page.web else True,
                    on_click=lambda e: page.show_dialog(
                        settings_dlg(handle_open_location_settings)
                    ),
                ),
                ft.OutlinedButton(
                    "open_app_settings",
                    visible=False if page.web else True,
                    on_click=lambda e: page.show_dialog(
                        settings_dlg(handle_open_app_settings)
                    ),
                ),
            ],
            wrap=True,
        )
    )


ft.app(main)
```

File: packages/flet_geolocator/lib/src/geolocator.dart (hunk 5 above)
Comment (ndonkoHenri) [reply packages/flet_geolocator/lib/src/geolocator.dart]: No worries, I did this.
Marking conversation as resolved.

File: client/android/app/build.gradle [hunk 6]
```diff
@@ -25,7 +25,7 @@ if (flutterVersionName == null) {
 android {
     namespace "com.appveyor.flet"
     compileSdkVersion flutter.compileSdkVersion
-    ndkVersion flutter.ndkVersion
+    ndkVersion "25.1.8937393"
```
Comment (FeodorFitsner) [inline client/android/app/build.gradle:28]: Is it accidental change or this is required by Geolocation plugin?

File: packages/flet_geolocator/pubspec.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  sdk: '>=3.2.3 <4.0.0'
+  flutter: ">=1.17.0"
+
+dependencies:
+  flutter:
+    sdk: flutter
+
+  collection: ^1.16.0
+  geolocator: ^11.0.0
+  permission_handler: ^11.3.1
```
Comment (FeodorFitsner) [inline packages/flet_geolocator/pubspec.yaml:17]: I guess we could remove it as it's not used anywhere in the code.

File: client/android/app/build.gradle (hunk 6 above)
Comment (lekshmanmj) [reply client/android/app/build.gradle:28]: It's required by flutter while testing on android device

**Review by FeodorFitsner — APPROVED** [review]

**Comment by Saidjony** [comment]
image

help

**Comment by Kacper-OS** [comment]
> image
> 
> help

`pip install flet --upgrade --pre`

**Pull request merged**